### PR TITLE
TFFR1 - Category: Composition - Cross-Referencing Modules

### PR DESCRIPTION
**Pull request by neil-yechenwei**
--- PASS: TestModuleSource/source_exists,_ok (0.00s)
    --- PASS: TestModuleSource/git_reference,_not_ok (0.00s)
    --- PASS: TestModuleSource/github_reference,_not_ok (0.00s)
    --- PASS: TestModuleSource/no_source,_not_ok (0.00s)
    --- PASS: TestModuleSource/no_version,_ok (0.00s)

### CODE REVIEW

File: integration/optional-defaults-incorrect/terraform.tf [hunk 1]
```diff
@@ -0,0 +1,13 @@
+terraform {
+  required_version = "~> 1.7.0"
+  required_providers {
+    azurerm = {
+      source  = "hashicorp/azurerm"
+      version = ">= 3.97.0, < 3.99.0"
```
Comment (lonegunmanb) [inline integration/optional-defaults-incorrect/terraform.tf:6]: Why not `">= 3.97.0, < 4.0.0"`？

File: integration/optional-defaults/terraform.tf [hunk 2]
```diff
@@ -0,0 +1,13 @@
+terraform {
+  required_version = "~> 1.7.0"
+  required_providers {
+    azurerm = {
+      source  = "hashicorp/azurerm"
+      version = ">= 3.97.0, < 3.99.0"
```
Comment (lonegunmanb) [inline integration/optional-defaults/terraform.tf:6]: Same issue here

File: integration/simplevaluerule-null-value/terraform.tf [hunk 3]
```diff
@@ -0,0 +1,13 @@
+terraform {
+  required_version = "~> 1.7.0"
+  required_providers {
+    azurerm = {
+      source  = "hashicorp/azurerm"
+      version = ">= 3.97.0, < 3.99.0"
```
Comment (lonegunmanb) [inline integration/simplevaluerule-null-value/terraform.tf:6]: Same issue here

File: integration/unknownrule-null-incorrect/terraform.tf [hunk 4]
```diff
@@ -0,0 +1,13 @@
+terraform {
+  required_version = "~> 1.7.0"
+  required_providers {
+    azurerm = {
+      source  = "hashicorp/azurerm"
+      version = ">= 3.97.0, < 3.99.0"
```
Comment (lonegunmanb) [inline integration/unknownrule-null-incorrect/terraform.tf:6]: Same issue here

File: integration/unknownrule-null/terraform.tf [hunk 5]
```diff
@@ -0,0 +1,13 @@
+terraform {
+  required_version = "~> 1.7.0"
+  required_providers {
+    azurerm = {
+      source  = "hashicorp/azurerm"
+      version = ">= 3.97.0, < 3.99.0"
```
Comment (lonegunmanb) [inline integration/unknownrule-null/terraform.tf:6]: Same issue here

File: rules/modules.go [hunk 6]
```diff
@@ -0,0 +1,92 @@
+package rules
```
Comment (lonegunmanb) [inline rules/modules.go:1]: I'd like to rename `modules.go` to `module_source.go`.

File: rules/modules.go [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/hashicorp/hcl/v2"
+	"github.com/hashicorp/hcl/v2/hclsyntax"
+	"github.com/terraform-linters/tflint-plugin-sdk/tflint"
+)
+
+var _ tflint.Rule = new(ModulesRule)
+
+type ModulesRule struct {
```
Comment (lonegunmanb) [inline rules/modules.go:14]: Could we rename it to `ModuleSourceRule`?

File: rules/modules.go [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+	tflint.DefaultRule
+}
+
+func NewModulesRule() *ModulesRule {
+	return &ModulesRule{}
+}
+
+func (t *ModulesRule) Name() string {
+	return "required_providers_tffr1"
```
Comment (lonegunmanb) [inline rules/modules.go:23]: The name could be `required_module_source_tffr1`.

**Review by lonegunmanb — CHANGES_REQUESTED** [review]
Thanks @neil-yechenwei for opening this pr! Some comments.

File: rules/modules.go (hunk 8 above)
Comment (neil-yechenwei) [reply rules/modules.go]: Updated

File: rules/modules.go (hunk 7 above)
Comment (neil-yechenwei) [reply rules/modules.go]: Updated

File: rules/modules.go (hunk 6 above)
Comment (neil-yechenwei) [reply rules/modules.go]: Updated

File: integration/unknownrule-null/terraform.tf (hunk 5 above)
Comment (neil-yechenwei) [reply integration/unknownrule-null/terraform.tf]: Updated

File: integration/simplevaluerule-null-value/terraform.tf (hunk 3 above)
Comment (neil-yechenwei) [reply integration/simplevaluerule-null-value/terraform.tf]: Updated

File: integration/optional-defaults/terraform.tf (hunk 2 above)
Comment (neil-yechenwei) [reply integration/optional-defaults/terraform.tf]: Updated

File: integration/optional-defaults-incorrect/terraform.tf (hunk 1 above)
Comment (neil-yechenwei) [reply integration/optional-defaults-incorrect/terraform.tf]: Updated

**Comment by neil-yechenwei** [comment]
@lonegunmanb , thanks for the comments. I updated PR. Please take another look. Thanks.

**Review by lonegunmanb — APPROVED** [review]
LGTM! Thanks @neil-yechenwei , but I'd like to let @matt-FFFFFF have a review too.

File: rules/module_source.go [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+	if err := r.EvaluateExpr(source.Expr, t.isAVMModule(r, source.NameRange), &tflint.EvaluateExprOption{ModuleCtx: tflint.RootModuleCtxType}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (t *ModuleSourceRule) isAVMModule(r tflint.Runner, issueRange hcl.Range) func(string) error {
+	return func(source string) error {
+		if !(strings.HasPrefix(source, "Azure/") && strings.HasSuffix(source, "/azurerm")) && !strings.HasPrefix(source, "..") {
```
[the remark this answers is not in the document]
Comment (neil-yechenwei) [reply rules/module_source.go]: Updated

**Comment by neil-yechenwei** [comment]
@matt-FFFFFF & @lonegunmanb , thanks for the comments. I updated PR. Please take another look.

**Comment by lonegunmanb** [comment]
> Thinking about this...
> 
> Would it make sense to extend the simple attribute value package to allow for a function to be passed on to validate the value?
> 
> At the moment we compare the value to a supplied slice of values, but we could take a function with signature:
> 
> `type checkFunc func[T comparable](val T) (bool, error)`
> 
> There is no difference to check g a module block as there is a resource block?
> 
> WDYT @lonegunmanb

@matt-FFFFFF 
I think maybe we can accept this one, since the check on `source` is not a simple comparation. As for the check callback you mentioned, I like the idea, maybe we can open another refactor pr to do this.

**Pull request merged**